controllers/perconaservermysql_controller.go

```diff
@@ -99,19 +99,90 @@ func (r *PerconaServerMySQLReconciler) Reconcile(
 		return rr, errors.Wrap(err, "get CR")
 	}
 
+	if cr.ObjectMeta.DeletionTimestamp != nil {
+		return rr, r.applyFinalizers(ctx, cr)
+	}
+
 	defer func() {
 		if err := r.reconcileCRStatus(ctx, cr); err != nil {
 			l.Error(err, "failed to update status")
 		}
 	}()
-
+	
 	if err := r.doReconcile(ctx, cr); err != nil {
 		return rr, errors.Wrap(err, "reconcile")
 	}
 
 	return rr, nil
 }
 
+func (r *PerconaServerMySQLReconciler) applyFinalizers(ctx context.Context, cr *apiv1alpha1.PerconaServerMySQL) error {
+	l := log.FromContext(ctx).WithName("Finalizer")
+
+	l.Info("Applying finalizers", "CR", cr)
+
+	var err error
+
+	finalizers := []string{}
+	for _, f := range cr.GetFinalizers() {
+		switch f {
+		case "delete-ps-pods-in-order":
+			l.Info("deleting PS pods")
+			err = r.deletePSPods(ctx, cr)
+		}
+
+		if err != nil {
+			l.Error(err, "failed to run finalizer", "finalizer", f)
+			finalizers = append(finalizers, f)
+		}
+	}
+
+	cr.SetFinalizers(finalizers)
+	err = r.Client.Update(ctx, cr)
+	if err != nil {
+		l.Error(err, "Client.Update failed")
+	}
+
+	return err
+}
+
+func (r *PerconaServerMySQLReconciler) deletePSPods(ctx context.Context, cr *apiv1alpha1.PerconaServerMySQL) error {
+	l := log.FromContext(ctx).WithName("Finalizer")
+
+	pods, err := k8s.PodsByLabels(ctx, r.Client, mysql.MatchLabels(cr))
+	if err != nil {
+		l.Error(err, "failed to get the pods")
+		return errors.Wrap(err, "get pods")
+	}
+	l.Info("got pods", "pods", len(pods), "pod name", pods[0].GetName())
+
+	// the last pod left - we can leave it for the stateful set
+	if len(pods) <= 1 {
+		l.Info("one or less then one pod")
+		time.Sleep(time.Second * 3)
+		return nil
+	}
+
+	sts := &appsv1.StatefulSet{}
+	if err := r.Client.Get(ctx, mysql.NamespacedName(cr), sts); err != nil {
+		return errors.Wrap(err, "get MySQL statefulset")
+	}
+	l.Info("got statefulset", "sts", sts, "spec", sts.Spec)
+
+	if sts.Spec.Replicas == nil || *sts.Spec.Replicas != 1 {
+		dscaleTo := int32(1)
+		sts.Spec.Replicas = &dscaleTo
+		err = r.Client.Update(ctx, sts)
+		if err != nil {
+			l.Error(err, "failed to update STS")
+			return errors.Wrap(err, "downscale StatefulSet")
+		}
+		l.Info("sts replicaset downscaled", "sts", sts)
+	}
```

Comment:
please check PSMDB operator for the logic: https://github.com/percona/percona-server-mongodb-operator/blob/4fccceafc8e84cb0b9f0ded143fd9d2406539862/pkg/controller/perconaservermongodb/finalizers.go#L39-L76

we shouldn't remove the finalizer before pods are deleted